magda-io/magda
# Added authorisation to links through the single-record (records/id) API

### What this PR does

Merge #2762 first! This includes it so that int tests would work.

Fixes #2368 

This adds authorization to records that link to other records. So if a record has an aspect with a link defined in the schema, this restricts the records that come through that link - so if you don't have access to record A, you can't see it or become aware of its existence, even if you do have record B that links to it. This carries whether `dereference` is true or false - if dereference is false, you shouldn't even be able to see the ID of linked record, if dereference is true you shouldn't be able to see the inner record (unless you have access).

### To test
There's a test env up at https://issues-2368-single-record-dereference-auth.dev.magda.io, which I've pre-populated with some data.

There are four inner records - `inner-dataset-access-control-1` and `inner-dataset-access-control-2`, which have `object.registry.record.owner_orgunit` as their policy, and `inner-dataset-access-public-1` and `inner-dataset-access-public-2`, which have `object.registry.record.public` as their policy.

Then there's records that link to them. These either have `aspect-with-link-single` and `aspect-with-link-multiple` aspects, which have a single link and an array link respectively.

The records I've made are:
- `outer-public-inner-single-public` (`aspect=aspect-with-link-single`): you should be able to see both inner and outer datasets regardless of being logged in or out
- `outer-public-inner-single-control` (`aspect=aspect-with-link-single`): logged out, you should only be able to see the outer dataset, logged in as admin you should be able to see both inner and outer
- `outer-control-inner-single-public` (`aspect=aspect-with-link-single`): logged out, you shouldn't be able to see anything, logged in as admin you should be able to see inner and outer
- `outer-public-multiple` (`aspect=aspect-with-link-multiple`): logged out, you should be able to see two inner records, logged in as admin you should see all 4
- `outer-control-multiple` (`aspect=aspect-with-link-multiple`): logged out, you should see nothing, logged in as admin you should see four inner datasets

### Checklist
